icu4c/source/common/uloc.cpp

```diff
@@ -1184,104 +1184,128 @@ _getLanguage(const char* localeID,
-    if(_isIDPrefix(localeID)) {
-        result.append((char)uprv_tolower(*localeID), status);
-        result.append('-', status);
-        localeID+=2;
+    int32_t len = _isIDPrefix(localeID) ? 2 : 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len])) {
+        if (len == MAXLEN) {
+            status = U_ILLEGAL_ARGUMENT_ERROR;
+            return;
+        }
+        len++;
     }
 
-    /* copy the language as far as possible and count its length */
-    while(!_isTerminator(*localeID) && !_isIDSeparator(*localeID)) {
-        result.append((char)uprv_tolower(*localeID), status);
-        localeID++;
+    *pEnd = localeID + len;
+    if (sink == nullptr || len == 0) { return; }
+
+    int32_t minCapacity = uprv_max(len, 4);  // Minimum 3 letters plus NUL.
+    char scratch[MAXLEN];
+    int32_t capacity = 0;
+    char* buffer = sink->GetAppendBuffer(
+            minCapacity, minCapacity, scratch, UPRV_LENGTHOF(scratch), &capacity);
+
+    for (int32_t i = 0; i < len; ++i) {
+        buffer[i] = uprv_tolower(localeID[i]);
+    }
+    if (_isIDSeparator(localeID[1])) {
+        buffer[1] = '-';
     }
 
-    if(result.length()==3) {
+    if (len == 3) {
         /* convert 3 character code to 2 character code if possible *CWB*/
-        int32_t offset = _findIndex(LANGUAGES_3, result.data());
+        U_ASSERT(capacity >= 4);
+        buffer[3] = '\0';
+        int32_t offset = _findIndex(LANGUAGES_3, buffer);
         if(offset>=0) {
-            result.clear();
-            result.append(LANGUAGES[offset], status);
+            const char* const alias = LANGUAGES[offset];
+            sink->Append(alias, (int32_t)uprv_strlen(alias));
+            return;
         }
     }
 
-    if (sink != nullptr && !result.isEmpty()) {
-        sink->Append(result.data(), result.length());
-    }
-
-    *pEnd = localeID;
+    sink->Append(buffer, len);
 }
 
 static void
 _getScript(const char* localeID,
            ByteSink* sink,
-           const char** pEnd,
-           UErrorCode& status) {
+           const char** pEnd) {
     U_ASSERT(pEnd != nullptr);
-    CharString result;
-    int32_t idLen = 0;
-
     *pEnd = localeID;
 
-    /* copy the second item as far as possible and count its length */
-    while(!_isTerminator(localeID[idLen]) && !_isIDSeparator(localeID[idLen])
-            && uprv_isASCIILetter(localeID[idLen])) {
-        idLen++;
-    }
+    constexpr int32_t LENGTH = 4;
 
-    /* If it's exactly 4 characters long, then it's a script and not a country. */
-    if (idLen == 4) {
-        int32_t i;
-        *pEnd = localeID + idLen;
-        if (idLen >= 1) {
-            result.append((char)uprv_toupper(*(localeID++)), status);
-        }
-        for (i = 1; i < idLen; i++) {
-            result.append((char)uprv_tolower(*(localeID++)), status);
-        }
+    int32_t len = 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len]) &&
```

Comment:
optional: we can probably simply check for letters, since terminators and separators are not letters:
```
while (uprv_isASCIILetter(localeID[len])) {
```